flang/test/Driver/linker-flags.f90

```diff
@@ -4,6 +4,7 @@
 
 ! RUN: %flang -### -target ppc64le-linux-gnu %S/Inputs/hello.f90 2>&1 | FileCheck %s --check-prefixes=CHECK,GNU
 ! RUN: %flang -### -target aarch64-apple-darwin %S/Inputs/hello.f90 2>&1 | FileCheck %s --check-prefixes=CHECK,DARWIN
+! RUN: %flang -### -target sparc-sun-solaris2.11 %S/Inputs/hello.f90 2>&1 | FileCheck %s --check-prefixes=CHECK,GNU
```

Comment:
Solaris is not GNU ;-) Please use a dedicated prefix.

Reply:
True, but how many more copies of the same patterns do we want?  Add FreeBSD, OpenBSD, ... whenever they arrive?  Isn't the common pattern <linker> <object file> <flang rt libs> with or without `-lm`?  How about putting the first under one label and the `-lm` part under another, used where appropriate?

Reply:
Sure, reducing duplication would be nice. Could you propose something?

Reply:
@banach-space How should we proceed with this one?  Really introduce separate labels for each OS, even if the checks are identical?  Or consolidate labels as I suggested?

Reply:
I would consolidate like you suggested. Perhaps the common prefix cold be "ALL-RT-LIBS" or "FLANG-RT-LIBS"? Suggestions welcome.

Reply:
> Add FreeBSD, OpenBSD, ... whenever they arrive?

Bumping into this PR that'll be happening much sooner for all of the *BSD's now.

Reply:
One thing to add to this. I noticed the Gnu addition does not check for `nostdlib, nodefaultlibs, or r` flags like the C++ library chunk does. With the proposed diff and what I had initially being right under the C++ chunk in the Solaris and *BSD ToolChain they would be checking for said flags.

Reply:
@brad0 , perhaps I misunderstood your comment, but "linker-flags.f90"checks for things specific to LLVM Flang. Generic stuff that applies to C++ and Fortran should be tested in Clang (i.e. where `clangDriver` library is located).

Reply:
> @brad0 , perhaps I misunderstood your comment, but "linker-flags.f90"checks for things specific to LLVM Flang. Generic stuff that applies to C++ and Fortran should be tested in Clang (i.e. where `clangDriver` library is located).

My comment was not about the test. Because of where the chunk is being added in the Solaris ToolChain (and what I had in my tree for *BSD's) it is checking for the ```nostdlib, nodefaultlibs, or r``` flags, where as the Gnu path does not. I just want to make sure the behavior is consistent.

tools::gnutools::Linker::ConstructJob

```
  if (D.CCCIsCXX() &&
      !Args.hasArg(options::OPT_nostdlib, options::OPT_nodefaultlibs,
                   options::OPT_r)) {
    if (ToolChain.ShouldLinkCXXStdlib(Args)) {
      bool OnlyLibstdcxxStatic = Args.hasArg(options::OPT_static_libstdcxx) &&
                                 !Args.hasArg(options::OPT_static);
      if (OnlyLibstdcxxStatic)
        CmdArgs.push_back("-Bstatic");
      ToolChain.AddCXXStdlibLibArgs(Args, CmdArgs);
      if (OnlyLibstdcxxStatic)
        CmdArgs.push_back("-Bdynamic");
    }
    CmdArgs.push_back("-lm");
  }

  // Silence warnings when linking C code with a C++ '-stdlib' argument.
  Args.ClaimAllArgs(options::OPT_stdlib_EQ);

  // Additional linker set-up and flags for Fortran. This is required in order
  // to generate executables. As Fortran runtime depends on the C runtime,
  // these dependencies need to be listed before the C runtime below (i.e.
  // AddRuntTimeLibs).
  if (D.IsFlangMode()) {
    addFortranRuntimeLibraryPath(ToolChain, Args, CmdArgs);
    addFortranRuntimeLibs(ToolChain, CmdArgs);
    CmdArgs.push_back("-lm");
  }
```

solaris::Linker::ConstructJob

```
  if (!Args.hasArg(options::OPT_nostdlib, options::OPT_nodefaultlibs,
                   options::OPT_r)) {
    if (D.CCCIsCXX()) {
      if (getToolChain().ShouldLinkCXXStdlib(Args))
        getToolChain().AddCXXStdlibLibArgs(Args, CmdArgs);
      CmdArgs.push_back("-lm");
    }
    // Additional linker set-up and flags for Fortran. This is required in order
    // to generate executables. As Fortran runtime depends on the C runtime,
    // these dependencies need to be listed before the C runtime below.
    if (D.IsFlangMode()) {
      addFortranRuntimeLibraryPath(getToolChain(), Args, CmdArgs);
      addFortranRuntimeLibs(getToolChain(), CmdArgs);
      CmdArgs.push_back("-lm");
    }
```

Reply:
One thing to add to this. I noticed the Gnu addition does not check for `nostdlib, nodefaultlibs, or r` flags like the C++ library chunk does. With the proposed diff and what I had initially being right under the C++ chunk in the Solaris and *BSD ToolChain they would be checking for said flags.